File: src/main/scala/com/abaddon83/burraco/match/games/domainModels/DiscardPile.scala
package com.abaddon83.burraco.`match`.games.domainModels

import com.abaddon83.burraco.shares.decks.{Card, Deck}

import scala.collection.mutable.ListBuffer

case class DiscardPile private(override protected val cards: ListBuffer[Card]) extends Deck{

  override def grabAllCards(): List[Card] ={
    assert(cards.size >0,"The DiscardPile is empty, you can't grab a card from here")
    super.grabAllCards()
  }

  override def grabFirstCard(): Card = {
    throw new UnsupportedOperationException("You cannot grab only one card from the DiscardPile")
  }

  def addCard(card: Card): DiscardPile = {
    cards.addOne(card)
    DiscardPile(cards)
  }
}

object DiscardPile{
  def apply(cards: List[Card]): DiscardPile = new DiscardPile(cards.to(ListBuffer))
}

File: src/main/kotlin/com/abaddon83/burracoGame/writeModel/models/DiscardPile.kt
package com.abaddon83.burracoGame.writeModel.models

import com.abaddon83.burracoGame.writeModel.models.decks.Card
import com.abaddon83.burracoGame.writeModel.models.decks.Deck
import com.abaddon83.utils.logs.WithLog

data class DiscardPile private constructor(override val cards: MutableList<Card>): Deck,WithLog() {

    override fun grabAllCards(): List<Card> {
        check(cards.size >0){warnMsg("The DiscardPile is empty, you can't grab a card from here")}
        return super.grabAllCards()
    }

    override fun grabFirstCard(): Card = throw UnsupportedOperationException("You cannot grab only one card from the DiscardPile")


    fun addCard(card: Card): DiscardPile {
        cards.add(card)
        return this
    }

    fun showCards(): List<Card> = cards.toList()

    companion object Factory {
        fun create(cards: List<Card>): DiscardPile {
            return DiscardPile(cards = cards.toMutableList())
        }
    }
}